distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/MachineDependentHeap.java

```diff
@@ -95,13 +99,13 @@ protected int getHeapSizeMb(Settings nodeSettings, MachineNodeRole role, long av
              *
              * If this formula is changed then corresponding changes must be made to the {@code NativeMemoryCalculator} and
              * {@code MlAutoscalingDeciderServiceTests} classes in the ML plugin code. Failure to keep the logic synchronized
-             * could result in repeated autoscaling up and down.
+             * could result in ML processes crashing with OOM errors or repeated autoscaling up and down.
              */
             case ML_ONLY -> {
                 if (availableMemory <= (GB * 16)) {
-                    yield mb((long) (availableMemory * .4), 4);
+                    yield mb((long) (availableMemory * .4 * 2 / 3), 4);
```

Comment:
The implicit dependency on the JavaErgonomics code to always give 1/2 of the `heapSize` to `MaxDirectMemorySize` seems fragile to me. It would be better to extract the code 
`ergonomicChoices.add("-XX:MaxDirectMemorySize=" + heapSize / 2);` from JavaErgonomics into a function and then use this function explicitly here to determine the memory or at least to log error if this assumption is broken.